### PR TITLE
Namespace restriction: revised

### PR DESCRIPTION
**Pull request by struz**
Update of https://github.com/jtblin/kube2iam/pull/39 to get it working. Only minor tweaks based on conflicts and merges that happened since @SleepyBrett last committed an update.

Apologies about the merge spam commits - hindsight says I should have rebased.

### CODE REVIEW

File: cmd/k8s.go [hunk 1]
```diff
@@ -40,6 +40,26 @@ func (k8s *k8s) watchForPods(podManager kcache.ResourceEventHandler) kcache.Stor
...
+func (k8s *k8s) watchForNamespaces(nsManager kcache.ResourceEventHandler) kcache.Store {
+	nsStore, nsController := kcache.NewInformer(
+		k8s.createNamespaceLW(),
+		&api.Namespace{},
+		resyncPeriod,
+		kcache.ResourceEventHandlerFuncs{
+			AddFunc:    nsManager.OnAdd,
+			DeleteFunc: nsManager.OnDelete,
+			UpdateFunc: nsManager.OnUpdate,
+		},
```
Comment (ash2k) [inline cmd/k8s.go]: Just pass `nsManager` directly, no need to wrap it in another layer of indirection.

File: cmd/namespace.go [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+	ns, ok := obj.(*api.Namespace)
+	if !ok {
+		log.Errorf("Expected Namespace but OnAdd handler received %+v", obj)
+		return
+	}
+
+	log.Debugf("Namespace OnAdd %s", ns.GetName())
+
+	roles := h.getRoleAnnotation(ns)
+	if roles != nil {
```
Comment (ash2k) [inline cmd/namespace.go]: This check is unnecessary.

File: cmd/namespace.go [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+		for i := range roles {
```
Comment (ash2k) [inline cmd/namespace.go]: `for _, role := ....` and use `role` in the loop body

File: cmd/namespace.go [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+	"k8s.io/kubernetes/pkg/api"
+)
+
+type namespacehandler struct {
+	storage *store
+}
+
+// OnAdd called with a namespace is added to k8s
+func (h *namespacehandler) OnAdd(obj interface{}) {
+	ns, ok := obj.(*api.Namespace)
```
Comment (ash2k) [inline cmd/namespace.go]: I think it is ok to just cast "dangerously" - if some other type of object has been passed that is a serious bug somewhere and program should crash.

File: cmd/server.go [hunk 5]
```diff
@@ -79,6 +82,23 @@ func (s *Server) getRole(IP string) (string, error) {
...
+func (s *Server) debugStoreHandler(w http.ResponseWriter, r *http.Request) {
+	output := make(map[string]interface{})
+
+	output["rolesByIP"] = s.store.DumpRolesByIP()
+	output["rolesByNamespace"] = s.store.DumpRolesByNamespace()
+	output["namespaceByIP"] = s.store.DumpNamespaceByIP()
+
+	o, err := json.Marshal(output)
+	if err != nil {
+		log.Errorf("Error converting debug map to json: %+v", err)
```
Comment (ash2k) [inline cmd/server.go]: This block should also return HTTP 500 instead of falling through and writing uninitialized `o` and returning 200 then.

File: cmd/server.go [hunk 6]
```diff
@@ -79,6 +82,23 @@ func (s *Server) getRole(IP string) (string, error) {
...
+	}
+
+	if _, err := w.Write(o); err != nil {
+		log.Errorf("Error writing response: %+v", err)
```
Comment (ash2k) [inline cmd/server.go]: This is unnecessary - most likely client has just disconnected and there is no point logging that. Not on `error` in any case.

File: cmd/server.go [hunk 7]
```diff
@@ -99,17 +119,24 @@ func (s *Server) securityCredentialsHandler(w http.ResponseWriter, r *http.Reque
...
-	allowedRoleARN := s.iam.roleARN(allowedRole)
+	podRole, err := s.getRole(remoteIP)
 	if err != nil {
 		http.Error(w, err.Error(), http.StatusNotFound)
 		return
 	}
+	podRoleARN := s.iam.roleARN(podRole)
+
+	if !s.store.CheckNamespaceRestriction(podRoleARN, remoteIP) {
+		http.Error(w, fmt.Sprintf("Role requested %s not valid for namespace of pod at %s", podRole, remoteIP), http.StatusNotFound)
```
Comment (ash2k) [inline cmd/server.go]: Maybe put pod name instead of ip? and also mention namespace

File: cmd/server.go [hunk 8]
```diff
@@ -119,7 +146,7 @@ func (s *Server) roleHandler(w http.ResponseWriter, r *http.Request) {
 
 	credentials, err := s.iam.assumeRole(wantedRoleARN, remoteIP)
 	if err != nil {
-		log.Errorf("Error assuming role %+v", err)
+		log.Errorf("Error assuming role %+v for pod at %s", err, remoteIP)
```
Comment (ash2k) [inline cmd/server.go]: Pod name+namespace?

File: cmd/store.go [hunk 9]
```diff
@@ -31,64 +35,132 @@ func (s *store) Get(IP string) (string, error) {
...
+			c = false
+			break
 		}
 	}
+	if c {
+		ar = append(ar, roleARN)
+	}
+	s.mutex.Lock()
+	s.rolesByNamespace[namespace] = ar
+	s.mutex.Unlock()
```
Comment (ash2k) [inline cmd/store.go]: Why all mutex unlocks in this file are not using `defer`?

File: cmd/store.go [hunk 10]
```diff
@@ -31,64 +35,132 @@ func (s *store) Get(IP string) (string, error) {
...
-		defaultRole: defaultRole,
-		iamRoleKey:  key,
-		rolesByIP:   make(map[string]string),
+		defaultRole:          defaultRole,
+		iamRoleKey:           key,
+		namespaceKey:         namespaceKey,
+		namespaceRestriction: namespaceRestriction,
+		rolesByIP:            make(map[string]string),
+		rolesByNamespace:     make(map[string][]string),
+		namespaceByIP:        make(map[string]string),
```
Comment (ash2k) [inline cmd/store.go]: I think the whole store should not exist. It should be implementable via IndexedInformers with custom indexes for these ^ things. Then you don't need to maintain these mappings yourself. However this may be done in a separate PR later if this works for you.

File: cmd/namespace.go (hunk 4 above)
Comment (ash2k) [inline cmd/namespace.go]: Same in other places.

File: cmd/namespace.go (hunk 2 above)
Comment (ash2k) [inline cmd/namespace.go]: Same in other places.

File: cmd/namespace.go (hunk 3 above)
Comment (ash2k) [inline cmd/namespace.go]: Same in other places.

File: cmd/namespace.go (hunk 4 above)
Comment (jtblin) [inline cmd/namespace.go]: I'd keep the cast check here for consistency with other places. We can change that later everywhere if we want to. Same as kubernetes code base btw.

File: cmd/store.go (hunk 10 above)
Comment (jtblin) [inline cmd/store.go]: Yeah there's another PR that was proposing that. But it would need serious rebasing and does not include namespaces obviously. Not sure why they don't use that in the kubernetes code base for DNS though and if there is any trade-off.

Comment (ash2k) [inline cmd/store.go]: Yeah, I saw that pr. It is a good idea. My guess would be because informers were not available back then. Informers are the right way to write controllers in most cases. See for more info https://github.com/kubernetes/community/blob/master/contributors/devel/controllers.md

Comment (jtblin) [inline cmd/store.go]: We already use informers (which the doc you link to seems to be about), the other PR and I assume your comment about the store are for IndexerInformer ;)

Anyhow, agreed. Let's finish this guy then we can move both the standard path and namespace to IndexerInformer.

File: cmd/server.go (hunk 7 above)
Comment (struz) [inline cmd/server.go]: We don't actually store/track the pod name anywhere. Perhaps we could do this in another PR.

File: cmd/server.go (hunk 8 above)
Comment (struz) [inline cmd/server.go]: See other comment regarding pod name. I will include the namespace though.

File: cmd/k8s.go (hunk 1 above)
Comment (struz) [inline cmd/k8s.go]: I'm not really sure what you mean here. All the examples I can find specify functions specifically, but I am probably missing something.

File: cmd/namespace.go (hunk 4 above)
Comment (struz) [inline cmd/namespace.go]: I'll keep it for now.

File: cmd/store.go (hunk 9 above)
Comment (struz) [inline cmd/store.go]: I think it's really just a style thing. As far as I can tell for simple things like this, explicit lock/unlock code will be ever so slightly more performant but in reality it's probably negligible. Happy to change this if people think it looks better, otherwise I'll stick with what was already there.

File: README.md [hunk 11]
```diff
@@ -191,6 +191,29 @@ spec:
...
+
+To allow the aws-cli pod specified above to run in the default namespace your namespace would look like the following.
+
+```
+---
+apiVersion: v1
+kind: Namespace
+metadata:
+  annotations:
+    kube2iam/allowed-roles: |
```
Comment (jtblin) [inline README.md]: Shouldn't we use something more consistent with the non namespace option e.g. `iam.amazonaws.com/allowed-roles`

File: cmd/namespace.go [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+		return
+	}
+	log.Debugf("Namespace OnDelete %s", ns.GetName())
+	h.storage.DeleteNamespace(ns.GetName())
+}
+
+// getRoleAnnotations reads the "kube2iam/roles" annotation off a namespace
+// and splits them as a JSON list (["role1", "role2", "role3"])
+func (h *namespacehandler) getRoleAnnotation(ns *api.Namespace) []string {
+	rolesstring := ns.Annotations[h.storage.namespaceKey]
```
Comment (jtblin) [inline cmd/namespace.go]: Use camel case for "rolesstring" or just "roles".

**Comment by struz** [comment]
Okay I've updated the branch addressing the requested changes now.

File: cmd/k8s.go (hunk 1 above)
Comment (ash2k) [inline cmd/k8s.go]:
```
nsStore, nsController := kcache.NewInformer(
		k8s.createNamespaceLW(),
		&api.Namespace{},
		resyncPeriod,
		nsManager
)
```

Comment (ash2k) [inline cmd/k8s.go]: `NewInformer()` wants `ResourceEventHandler` as its last argument and `nsManager` already implements that interface so you can just pass it directly.

File: cmd/server.go [hunk 13]
```diff
@@ -79,6 +82,22 @@ func (s *Server) getRole(IP string) (string, error) {
...
+	output := make(map[string]interface{})
+
+	output["rolesByIP"] = s.store.DumpRolesByIP()
+	output["rolesByNamespace"] = s.store.DumpRolesByNamespace()
+	output["namespaceByIP"] = s.store.DumpNamespaceByIP()
+
+	o, err := json.Marshal(output)
+	if err != nil {
+		log.Errorf("Error converting debug map to json: %+v", err)
+		http.Error(w, err.Error(), http.StatusInternalServerError)
```
Comment (ash2k) [inline cmd/server.go]: Well, you need to exit the function after this line :)

**Comment by struz** [comment]
And again. Let me know if there's anything major outstanding.

File: cmd/store.go [hunk 14]
```diff
@@ -31,64 +35,133 @@ func (s *store) Get(IP string) (string, error) {
...
+
+// AddRoleToNamespace takes a role name and adds it to our internal state
+func (s *store) AddRoleToNamespace(namespace string, role string) {
+	// Make sure to add the full ARN of roles to ensure string matching works
+	roleARN := s.iam.roleARN(role)
+
+	ar := s.rolesByNamespace[namespace]
+	if ar == nil {
+		ar = []string{}
 	}
```
Comment (ash2k) [inline cmd/store.go]: This block is unnecessary - appending to an empty slice is ok - it will allocate memory/etc automatically.

File: cmd/store.go [hunk 15]
```diff
@@ -31,64 +35,133 @@ func (s *store) Get(IP string) (string, error) {
...
-		if dok {
-			pod, ok = deletedObj.Obj.(*api.Pod)
+// checkRoleForNamespace checks the 'database' for a role allowed in a namespace,
+// returns true if the role is found, otheriwse false
+func (s *store) checkRoleForNamespace(role string, namespace string) bool {
+	ar := s.rolesByNamespace[namespace]
+	if ar == nil {
+		log.Warnf("Role:%s on namespace:%s not found.", role, namespace)
+		return false
+	}
```
Comment (ash2k) [inline cmd/store.go]: This `if` block is unnecessary, can be just deleted.

File: cmd/store.go [hunk 16]
```diff
@@ -31,64 +35,133 @@ func (s *store) Get(IP string) (string, error) {
...
+	for i := range ar {
```
Comment (ash2k) [inline cmd/store.go]: `for _, r := range ar` then `if r == role`.

**Comment by ash2k** [comment]
I left a few more comments, but no issues so I approved the PR 👍
